Remote host stderr should also append to $NVIM_LOG_FILE

phodge opened the pull request:
I've been trying to write a Node remote plugin and it's super frustrating because sometimes the remote host logs errors to `$NVIM_NODE_LOG_FILE`, but some runtime errors aren't caught and cause the node process to die, and the traceback isn't visible anywhere.

This PR modifies the call to `rpcstart()` so that if `$NVIM_<HOST>_LOG_FILE` is set, the remote host provider will attempt to write all of the process' stderr to `$NVIM_<HOST>_LOG_FILE` as well.

phodge committed:
provider.vim: just always send stderr to nvim_log()

justinmk commented:
a basic test would be good.

phodge commented on runtime/autoload/provider.vim:
```diff
@@ -19,3 +33,15 @@ function! provider#Poll(argv, orig_name, log_env) abort
+" Attempts to append the lines in 'data' to file at 'log_path'
+"
+" returns true if the file was written to successfully
+function! s:WriteLines(orig_name, log_path, job, data, event)
```
I'm actually somewhat concerned about how we're turning on capturing of stderr and never garbage collecting it. Once startup has been successful, the only thing the stderr stream is doing is leaking memory.

I'll try to rework the code a little so that we get the same message on crash but don't leak memory afterward.

bfredl replied:
Hosts are not supposed to write to stderr after startup is successful, rather there is methods for the host to send messages over the API, so there should be any memory leak going on continuously, only a message getting saved and immediately discarded on a non-startup crash ("garbage collection" already happens by then). Rather than disabling stderr collection after startup (which could be done just by adding `chanclose(job, 'stderr')` ), it might be useful to show the stderr also at a later crash (or at least stash it in `:messages` with no immediate display if it is too jarring)

phodge commented:
I've run into some challenges trying to implement this feature for the Python remote plugin hosts. The python-client has been "cleverly" designed such that instead of logging to `$NVIM_PYTHON_LOG_FILE`, it logs to a suffixed filename such as `$NVIM_PYTHON_LOG_FILE."_py3_rplugin"`. This means if I want `provider.vim` to write stderr to the same suffixed file, I need to duplicate the logic from python-client which generates the suffix, _and_ try to rework the remote/host/provider plugins such that they have enough context to figure out which suffix to mention in the startup error message. In fact, I'll need to duplicate the log-splitting logic _just_ so I can modify the startup error message such that it doesn't mislead users about where to find the python logs.

Personally I'm strongly in favour of getting rid of the log-suffixing logic from python-client and having all of the python hosts just write to `$NVIM_PYTHON_LOG_FILE`. That extra complexity is going to leach out into the provider code, and it's not even nice for our users anyway. I wasted a great deal of time using `:split $NVIM_PYTHON_LOG_FILE` to reload the log file and wondering why nothing was being logged before I stumbled across the suffixed file by accident. We could easily just prefix the log entries with the python version and type of host and users can `tail -f $NVIM_PYTHON_LOG_FILE | grep <whichever>` in the unlikely even they're getting too much noise from a python host they don't care about.

If we really must have separate files for each of the python hosts, then we should have 4 different variables for them so that our software is at least putting the logs where the user asked.

---

In summary, the options are:

1) Get rid of the log-suffixing logic from python-client and instead just put a prefix on the log entries.

2) Re-implement the log-suffixing logic in the remote host / provider plugins so that stderr is always logged to the same file that python-client is logging to.

3) Use 4 separate $NVIM_PYTHON<VERSION>_<KIND>_LOG_FILE` variables for the various python hosts.

bfredl commented:
> The python-client has been "cleverly" designed such that instead of logging to 

It was "cleverly" designed so the log file won't get overwritten by multiple versions of the python host, which would happen when one debugs in a realistic environment.

Why should neovim itself write anything at all to `$NVIM_PYTHON_LOG_FILE`? The file is meant for the python host to write to. Why can't the autoload provider just write to nvim's own log file?

justinmk commented:
In my mind, host-specific log files are deprecated in general. I'd prefer if all logging (except severe cases) fan-into the `nvim` API and everything goes to `$NVIM_LOG_FILE` or whatever sink `nvim` decides to write to. Currently this idea is vaporware in https://github.com/neovim/neovim/pull/7062 but if it makes sense for the friction described here, perhaps it won't take long to do a basic implementation.

phodge commented:
Writing absolutely everything to `$NVIM_LOG_FILE` sounds like the better solution to me as well, so that users will always know where to look for errors and tracebacks. If there's no objections, I'll add an `nvim_log()` API function (as per #7062) to this PR and push all the host stderr streams through it.

phodge committed:
tests: unit tests for api nvim_log()

phodge committed:
API: helper to get the name of an ObjectType

phodge committed:
logging: function to log a VimL array of lines

phodge committed:
logging: exract date/time population into a function

phodge committed:
logging: also sanitize the user-provided log_level

phodge committed:
API: nvim_log()

phodge committed:
DEV STASH: job labels

phodge committed:
wrap 'who' in brackets for better readability

phodge committed:
tests: clear() mustn't discard $VIMRUNTIME when overring env vars

phodge committed:
provider.vim: improve logging and error reporting

phodge committed:
provider.vim: provide a meaningful name for diagnostics

phodge committed:
provider.vim: basic unit tests

phodge commented:
Sorry for the delay guys, I had to rebuild an actual kitchen in the middle of all that.

I've removed the original 2 commits and replaced them with a bunch of new ones that do two things:

* Add an `nvim_log()` api function
* Modify `provider.vim` to send all stderr to `nvim_log()` (and also log a few more things at INFO level).

phodge committed:
tests: fix broken import

justinmk commented on src/nvim/log.c:
```diff
@@ -32,6 +32,25 @@ static uv_mutex_t mutex;
 
+#define LOCAL_TIME_STRING_LEN 20
+
+void get_local_time_string(char out[LOCAL_TIME_STRING_LEN])
```
I don't think we use array parameters anywhere. They can be misleading. Better to just use a pointer.

phodge replied:
I was trying to avoid having `get_local_time_string()` allocate a `char[]` that needs to be freed by the caller. Is that what you had in mind? Or would you have the caller allocate the `char[]` and pass the length as a 2nd argument? (I don't have enough C experience to know what the "normal" approach would be here, I'm assuming it's one of those two?)

jamessan replied:
While it's valid C syntax to declare the function as you did, it just gets collapsed down to a pointer.  Nothing enforces the length.

> I was trying to avoid having `get_local_time_string()` allocate a `char[]` that needs to be freed by the caller. Is that what you had in mind? Or would you have the caller allocate the `char[]` and pass the length as a 2nd argument?

It has to be one of the two. :) As long as it's clearly documented, either should be fine.  However, I would probably go with the latter.  Just make sure lengths are properly accounted for in `strftime()` and the fallback (`strlcpy()` or `strncpy()`).

justinmk commented on src/nvim/log.c:
```diff
@@ -137,6 +163,94 @@ bool logmsg(int log_level, const char *context, const char *func_name,
+      log_level_valid = (0 == strcmp(log_level, "ERROR")
+                         || 0 == strcmp(log_level, "WARN")
+                         || 0 == strcmp(log_level, "INFO")
+                         || 0 == strcmp(log_level, "DEBUG"));
```
we have `strequal()` FWIW

justinmk commented:
This looks pretty good, but is somewhat risky. WIill merge after 0.3.2 release.

phodge commented on runtime/autoload/provider.vim:
```diff
@@ -3,19 +3,46 @@
+endfun
+
+function! s:LogEvent(who, log_level, job, data, event)
+  call nvim_log(a:log_level, lines, {'who': who})
```
Hey that should be `data`, not `lines` :-S

phodge replied:
And even `a:data` + `a:who` for consistency.

phodge commented on src/nvim/log.c:
```diff
@@ -137,6 +163,94 @@ bool logmsg(int log_level, const char *context, const char *func_name,
 
+/// Log an array of lines to $NVIM_LOG_FILE. The only failure condition is that
+/// the log file can't be opened. Any other errors are written to the log file.
+bool do_log_array(char *log_level, Array lines, Dictionary opt)
```
The 3 options I considered were:

**A)** Create an array of `char *` with pointers into the VimL Array and pass that along to a generic array logging function in `vim.c`.

This is nice in that `log.c` doesn't need to know anything about VimL types, but then you have to send the error messages through in a 2nd call to the generic logging function, which introduces a race condition where another process may write to the log file in between any error lines and log lines. You can solve that by doing the locking/unlocking of the log file inside the `nvim_log()` API function and doing two consecutive calls to the generic logging function, but now you have a generic log function which relies on the caller to do the locking/unlocking, which feels untidy/unsafe.

**B)** Accept that we're going to have to lock/unlock the inside the `nvim_log()` API function, but instead of allocating potentially huge `char *` arrays up front, just iterate over the VimL array and call a generic line logging function once for each line.

I like that this solution avoids allocating the entire `char *` array up front, but don't like that it results in an extra function call for every single log line. This would mean regenerating the timestamp for every single line, or else you have to outsource *that* to the `nvim_log()` API function as well.

**C)** Just make a function in `log.c` that takes a VimL array.

This is a bit untidy in that `log.c` needs to start understanding VimL types, but it means that the locking, unlocking and timestamp generation stay within `log.c`, _and_ you don't have to allocate the `char *` array at all, _and_ the execution overhead of each log line is just the call to `fprintf()`.

---

The ability to handle unexpectedly huge data sets efficiently has always been one of the big selling points of Vim, so as much as possible I wanted this `nvim_log()` method to also cope well with unexpected huge log arrays, even though it's unlikely any client will ever send through anything more than maybe a 1000-line stack trace. That's why I picked option **C** above.

Thoughts?

phodge committed:
Merge branch 'master' into rplugin-does-logging

phodge committed:
log.c: use strequal()

phodge committed:
style

phodge committed:
log.c: Use xalloc()/xfree(); also fix a memory leak

phodge committed:
style

phodge committed:
log.c: rework call arguments for get_local_time_string()

phodge committed:
provider.vim: 'WARN' not 'WARNING'

phodge committed:
provider.vim: use correct variables

phodge committed:
test: prevent test passing by accident

phodge committed:
test: rework cleanup

phodge committed:
Merge branch 'master' into rplugin-does-logging

phodge committed:
Bump api version for nvim_log()

phodge committed:
Fix bad merge

bfredl commented:
#12503
